src/db_adapters/assets/non_fungible_token_events.rs

```diff
@@ -2,123 +2,72 @@ use actix_diesel::dsl::AsyncRunQueryDsl;
+            &detect_nft_db_error
         );
     }
+
     Ok(())
 }
 
-async fn is_error_handled(async_error: &AsyncError<diesel::result::Error>) -> bool {
-    if let actix_diesel::AsyncError::Execute(diesel::result::Error::DatabaseError(
-        diesel::result::DatabaseErrorKind::UniqueViolation,
-        ref error_info,
-    )) = *async_error
-    {
-        let duplicate_constraint = "assets__non_fungible_token_events_pkey";
-        let broken_data_constraint = "assets__non_fungible_token_events_unique";
-        let constraint_name = error_info.constraint_name().unwrap_or("");
-        if constraint_name == duplicate_constraint {
-            // Everything is fine, we have already written this to the DB
-            return true;
-        } else if constraint_name == broken_data_constraint {
-            warn!(
-                target: crate::INDEXER_FOR_EXPLORER,
-                "NFT: data inconsistency is found"
-            );
-        }
-    }
-    false
+async fn detect_nft_db_error(async_error: &AsyncError<diesel::result::Error>) -> bool {
+    assets::events::detect_db_error(
+        async_error,
+        "assets__non_fungible_token_events_pkey",
+        "assets__non_fungible_token_events_unique",
+    )
+    .await
 }
 
-fn collect_nft_events(
-    outcome: &near_indexer::IndexerExecutionOutcomeWithReceipt,
+fn compose_nft_db_events(
+    events_with_outcomes: &[(
+        assets::event_types::Nep171Event,
+        &near_indexer::IndexerExecutionOutcomeWithReceipt,
+    )],
     block_timestamp: u64,
     shard_id: &near_indexer::near_primitives::types::ShardId,
-    index_in_shard: &mut i32,
 ) -> Vec<models::assets::non_fungible_token_events::NonFungibleTokenEvent> {
-    let prefix = "EVENT_JSON:";
-    let event_logs = outcome.execution_outcome.outcome.logs.iter().filter_map(|untrimmed_log| {
-        // Now we have only nep171 events, we both parse the logs and handle nep171 here.
-        // When other event types will be added, we need to rewrite the logic
-        // so that we parse the logs only once for all,
-        // and then handle them for each event type separately.
-        let log = untrimmed_log.trim();
-        if !log.starts_with(prefix) {
-            return None;
-        }
-
-        let event: nft_types::NearEvent = match serde_json::from_str::<'_, nft_types::NearEvent>(
-            log[prefix.len()..].trim(),
-        ) {
-            Ok(result) => result,
-            Err(err) => {
-                warn!(
-                    target: crate::INDEXER_FOR_EXPLORER,
-                    "NFT: provided event log does not correspond to any of formats defined in NEP. Will ignore this event. \n {:#?} \n{:#?}",
-                    err,
-                    untrimmed_log,
-                );
-                return None;
-            }
-        };
-
-        let nft_types::NearEvent::Nep171(nep171_event) = event;
-        Some(nep171_event)
-    });
-
+    let mut index_in_shard = 0;
```

Comment:
We can replace it with `events_with_outcomes.enumerate()` on the loop

Reply:
No, we can't, since we have more than one DB event inside each log

Reply:
Oh, should we just use `ft_events.len()` instead of a separate variable `index_in_shard`?

Reply:
I'm not sure I got your idea.

We have 2 `for` cycles. We never calculate the sum of the sizes of the inner ones. We can add an additional cycle before this logic where we will calculate the capacity, but, IMO, that's overhead